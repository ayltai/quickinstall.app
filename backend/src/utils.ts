import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
