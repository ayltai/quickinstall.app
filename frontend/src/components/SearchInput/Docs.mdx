import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './SearchInput.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use search

Search lets users find information using keywords. Search can be used as a means of discovering content, or as a filter to help users in finding content.

When using search,
* Keep the original search query on the page so that user can easily edit it if they can't find what they are looking for

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Examples'
    includePrimary={false} />
