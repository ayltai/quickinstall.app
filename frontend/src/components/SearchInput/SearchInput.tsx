import SearchIcon from '@mui/icons-material/Search';
import CircularProgress from '@mui/material/CircularProgress';
import InputBase from '@mui/material/InputBase';
import { styled, } from '@mui/material/styles';
import React, { type ChangeEvent, type FC, useCallback, useEffect, } from 'react';
import { useDebounceValue, } from 'usehooks-ts';

import type { SearchInputProps, } from './SearchInput.types';

const Search = styled('div')({
    display         : 'inline-block',
    position        : 'relative',
    borderRadius    : '2rem',
    backgroundColor : 'rgba(0, 0, 0, 0.16)',
    '&:hover'       : {
        backgroundColor : 'rgba(0, 0, 0, 0.24)',
    },
});

const SearchIconWrapper = styled('div')(({ theme, }) => ({
    padding        : theme.spacing(0, 2),
    height         : '100%',
    position       : 'absolute',
    pointerEvents  : 'none',
    display        : 'flex',
    alignItems     : 'center',
    justifyContent : 'center',
}));

const StyledInputBase = styled(InputBase)(({ theme, }) => ({
    color                   : 'inherit',
    '& .MuiInputBase-input' : {
        padding                          : theme.spacing(1, 1, 1, 0),
        paddingLeft                      : `calc(1em + ${theme.spacing(4)})`,
        transition                       : theme.transitions.create('width'),
        width                            : '100%',
        [ theme.breakpoints.down('lg') ] : {
            width     : '20ch',
            '&:focus' : {
                width : '26ch',
            },
        },
        [ theme.breakpoints.up('lg') ]   : {
            width     : '48ch',
            '&:focus' : {
                width : '64ch',
            },
        },
    },
}));

/**
 * A search input lets users enter search queries and find relevant pieces of content.
 * @param loading Whether the input is loading
 * @param hint Placeholder text
 * @param onSearch Function to call when the input is changed
 * @param rest Other props
 */
export const SearchInput : FC<SearchInputProps> = ({
    loading,
    hint,
    onSearch,
    ...rest
}) => {
    const [ debouncedKeyword, setValue, ] = useDebounceValue('', 400);

    const handleChange = useCallback((event : ChangeEvent<HTMLInputElement>) => {
        setValue(event.target.value);
    }, [ setValue, ]);

    useEffect(() => {
        if (!loading && onSearch) onSearch(debouncedKeyword);
    }, [ debouncedKeyword, loading, onSearch, ]);

    return (
        <Search {...rest}>
            <SearchIconWrapper>
                {loading && (
                    <CircularProgress
                        size={24}
                        color='inherit' />
                )}
                {!loading && <SearchIcon />}
            </SearchIconWrapper>
            <StyledInputBase
                placeholder={hint}
                readOnly={loading}
                onChange={handleChange} />
        </Search>
    );
};
