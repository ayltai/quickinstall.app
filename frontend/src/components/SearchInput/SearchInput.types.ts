import type { HTMLAttributes, } from 'react';

export type SearchInputProps = HTMLAttributes<HTMLDivElement> & {
    loading?  : boolean,
    hint?     : string,
    onSearch? : (keyword? : string) => void,
};
