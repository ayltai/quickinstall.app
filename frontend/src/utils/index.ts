import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
