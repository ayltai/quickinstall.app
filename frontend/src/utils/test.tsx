import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
