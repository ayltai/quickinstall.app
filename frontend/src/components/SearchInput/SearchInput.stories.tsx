import type { Meta, StoryObj, } from '@storybook/react';

import { SearchInput, } from './SearchInput';

const meta : Meta<typeof SearchInput> = {
    component : SearchInput,
    title     : 'Components/Inputs/Search',
    argTypes  : {
        loading  : {
            description : 'Whether the input is loading',
        },
        hint     : {
            description : 'Placeholder text',
        },
        onSearch : {
            description : 'Function to call when the input is changed',
        },
    },
};

export default meta;

type Story = StoryObj<typeof SearchInput>;

export const Default : Story = {
    args : {
        hint     : 'Type something to search...',
        onSearch : () => {},
    },
};

export const Loading : Story = {
    args : {
        ...Default.args,
        loading : true,
    },
};
