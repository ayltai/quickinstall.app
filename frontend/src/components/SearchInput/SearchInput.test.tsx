import React from 'react';

import { fireEvent, render, waitFor, } from '../../utils/test';
import { SearchInput, } from './SearchInput';

describe('<SearchInput />', () => {
    it('renders correctly', () => expect(render(<SearchInput />)).toMatchSnapshot());

    it('triggers onSearch when the input is changed', async () => {
        const handleSearch = vi.fn();

        const { getByRole, } = render(<SearchInput onSearch={handleSearch} />);

        expect(handleSearch).toHaveBeenCalledTimes(1);

        fireEvent.change(getByRole('textbox'), {
            target : {
                value : 'test',
            },
        });

        await waitFor(() => {
            expect(handleSearch).toHaveBeenCalledTimes(2);
            expect(handleSearch).toHaveBeenCalledWith('test');
        });
    });
});
