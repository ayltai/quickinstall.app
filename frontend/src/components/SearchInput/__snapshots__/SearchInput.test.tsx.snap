// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchInput /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-1izsnok"
      >
        <div
          class="css-1iz4iln"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
            />
          </svg>
        </div>
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary css-8b2o5r-MuiInputBase-root"
        >
          <input
            class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1izsnok"
    >
      <div
        class="css-1iz4iln"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
          />
        </svg>
      </div>
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-8b2o5r-MuiInputBase-root"
      >
        <input
          class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
